src/WinRT.Runtime/ComWrappersSupport.net5.cs

```diff
@@ -212,51 +212,57 @@ static unsafe DefaultComWrappers()
         }
 
         protected override unsafe ComInterfaceEntry* ComputeVtables(object obj, CreateComInterfaceFlags flags, out int count)
-        {
-            if (IsRuntimeImplementedRCW(obj))
-            {
-                // If the object is a runtime-implemented RCW, let the runtime create a CCW.
-                count = 0;
-                return null;
-            }
-
-            var entries = ComWrappersSupport.GetInterfaceTableEntries(obj);
-
-            if (flags.HasFlag(CreateComInterfaceFlags.CallerDefinedIUnknown))
-            {
-                entries.Add(new ComInterfaceEntry
-                {
-                    IID = typeof(IUnknownVftbl).GUID,
-                    Vtable = IUnknownVftbl.AbiToProjectionVftblPtr
-                });
-            }
-
-            entries.Add(new ComInterfaceEntry
-            {
-                IID = typeof(IInspectable).GUID,
-                Vtable = IInspectable.Vftbl.AbiToProjectionVftablePtr
-            });
-
-            count = entries.Count;
-            ComInterfaceEntry* nativeEntries = (ComInterfaceEntry*)Marshal.AllocCoTaskMem(sizeof(ComInterfaceEntry) * count);
-
-            for (int i = 0; i < count; i++)
-            {
-                nativeEntries[i] = entries[i];
+        {
+            var vtableEntries = TypeVtableEntryTable.GetValue(obj.GetType(), (type) => 
+            {
+                if (IsRuntimeImplementedRCW(type))
+                {
+                    // If the object is a runtime-implemented RCW, let the runtime create a CCW.
+                    return new VtableEntries();
+                }
+
+                var entries = ComWrappersSupport.GetInterfaceTableEntries(type);
+
+                entries.Add(new ComInterfaceEntry
+                {
+                    IID = typeof(IInspectable).GUID,
+                    Vtable = IInspectable.Vftbl.AbiToProjectionVftablePtr
+                });
+
+                // This should be the last entry as it is included / excluded based on the flags.
+                entries.Add(new ComInterfaceEntry
+                {
+                    IID = typeof(IUnknownVftbl).GUID,
+                    Vtable = IUnknownVftbl.AbiToProjectionVftblPtr
+                });
+
+                ComInterfaceEntry* nativeEntries = (ComInterfaceEntry*)Marshal.AllocCoTaskMem(sizeof(ComInterfaceEntry) * entries.Count);

```

Comment:
I would avoid the manual managed allocation here and use [`RuntimeHelpers.AllocateTypeAssociatedMemory()`](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.runtimehelpers.allocatetypeassociatedmemory) instead.

Reply:
I did look at using it, but I didn't have a static field associated with the type that I could use to store it and I was concerned about storing this address in the ConditionalWeakTable as I wasn't sure whether there is a scenario where the memory allocated from this function might be cleaned up before the entry is removed / invalidated from the weak table.  If you think this latter case won't happen (i.e. I can't retrieve an entry from a ConditionalWeakTable based on type for which the type had already executed FreeMemory for any type associated memory), I can move to using it.

Reply:
>i.e. I can't retrieve an entry from a ConditionalWeakTable based on type for which the type had already executed FreeMemory for any type associated memory

Yep. This function was designed for exactly the scenario being performed here. The fix is simple - change the above call and remove the finalizer doing the cleanup in `VtableEntries`.

Reply:
This will also reduce pressure on the finalizer queue.